### ISSUE TITLE
Disallow staging of files with unresolved conflicts

### ISSUE THREAD

**Issue by Deilan** [Microsoft/vscode#30055]
Requesting to disallow staging of files with unresolved conflicts. Or at least show a warning message if a user is trying to stage a file with unresolved conflicts.

**Comment by joaomoreno** [comment]
This is a cool idea actually, want to attempt a PR?

**Comment by ferreus** [comment]
It's not that simple.
There are various types of conflicts possible, for example for conflict of types: (deleted by them / deleted by us) staging the file (git add), is a way to resole the conflict and keep the file.
In that case, refusing to stage the file, will prevent the user from resolving the conflict, and force the user to use alternative GIT client, to resole the conflict.

In such cases we can show a warning with: add/rm options. Or maybe show X,V buttons in the sidebar? But then, we need to show an indication of what actually happend , and a way to undo the operation before the actual commit. And it get's complicated.

For example, let's take this git conflict:
1. master modify file foo
2. branch fix delete file foo
3. merge fix to master
4. conflict: deleted by them

In this state, doing: git rm foo && git status
Shows:
Changes to be committed:
deleted foo
And this has a nice indication in VSCode

But if instead, i would do: git add foo && git status (Or stage the file from within VSCode)
I would see in terminal:
```
All conflicts fixed but you are still merging.
(use "git commit" to conclude merge)
```

Without any indication of our actions, and no way to undo the action.
VSCode shows the git repository as clean, without the need to commit at all. (Probably another issue)


I would be happy to work on this issue, but i don't know in what direction this should go.

**Comment by marckassay** [comment]
I see your concern.  What I observe in VSCode when there is a 'deleted by them' conflict on a file, in SCM viewlet under the 'Merge Changes' the file having a strike-through style on its name.  I find that there are 2 options:
- Upon mouse rollover, a '+' icon appears with the following tooltip: 'Stage Changes'.  If I click on this icon, Git will resolve the conflict which will make the branch updated with nothing to commit.  And that behavior is correct.  Git simply negated the deletion that "_by them_" did.   So that tooltip message is misleading, it didn’t stage the file, it simply resolved the conflict.
 
- If I don’t click on the '+' icon and attempt to commit, the following notification appears that prevents a successful commit:  'Error: Git: Committing is not possible because you have unmerged files.'

In a 'deleted by us' scenario, the difference when compared from 'deleted by them' scenario is that when the '+' icon is clicked, it is indeed staged as an new file (Index added).  So for the record, the only thing I see that needs to be fixed in a 'deleted by them' and 'deleted by us' scenario is the misleading tooltip message during a 'deleted by them' as mentioned above.

All other conflicts that contains conflict-markers (===, >>>, <<<) which are attempted to be added/staged should raise a warning as the OP suggested.

**Comment by BugraC** [comment]
@joaomoreno this is my first try. Please review when you have the chance #30568. Thank you.

**Comment by joaomoreno** [comment]
@BugraC @ferreus Let me introduce you guys to one another: #30568 #30529. 😆

**Comment by Deilan** [comment]
Sadly, both of the PRs couldn't pass CI. :(

**Comment by joaomoreno** [comment]
Fixed by #30568
